File: aoc2015/aoc2015-java/src/test/java/de/havox_design/aoc2015/day09/Day09Test.java
package de.havox_design.aoc2015.day09;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

class Day09Test {

    @ParameterizedTest
    @MethodSource("getDataForPart1")
    void testPart1(String fileName, int expectedShortestDistance) {
        Assertions.assertEquals(expectedShortestDistance, AllInASingleNight.solvePart1(fileName));
    }

    private static Stream<Arguments> getDataForPart1() {
        return Stream.of(
                Arguments.of("de/havox_design/aoc2015/day09/day09Sample.txt", 605)
        );
    }

    @ParameterizedTest
    @MethodSource("getDataForPart2")
    void testPart2(String fileName, int expectedLongestDistance) {
        Assertions.assertEquals(expectedLongestDistance, AllInASingleNight.solvePart2(fileName));
    }

    private static Stream<Arguments> getDataForPart2() {
        return Stream.of(
                Arguments.of("de/havox_design/aoc2015/day09/day09Sample.txt", 982)
        );
    }
}

File: aoc2023/day09/src/test/kotlin/de/havox_design/aoc2023/day09/Day09Test.kt
package de.havox_design.aoc2023.day09

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream

class Day09Test {
    @Test
    fun testMainClass() {
        MainClass.main(arrayOf())
    }

    @ParameterizedTest
    @MethodSource("getDataForTestSolvePart1")
    fun testSolvePart1(filename: String, expectedResult: Long) =
        MirageMaintenance(filename).solvePart1().shouldBe(expectedResult)

    @ParameterizedTest
    @MethodSource("getDataForTestSolvePart2")
    fun testSolvePart2(filename: String, expectedResult: Long) =
        MirageMaintenance(filename).solvePart2().shouldBe(expectedResult)

    companion object {
        @JvmStatic
        private fun getDataForTestSolvePart1(): Stream<Arguments> =
            Stream.of(
                Arguments.of("day09Sample1.txt", 18L),
                Arguments.of("day09Sample2.txt", 28L),
                Arguments.of("day09Sample3.txt", 68L),
                Arguments.of("day09Sample4.txt", 114L)
            )

        @JvmStatic
        private fun getDataForTestSolvePart2(): Stream<Arguments> =
            Stream.of(
                Arguments.of("day09Sample1.txt", -3L),
                Arguments.of("day09Sample2.txt", 0L),
                Arguments.of("day09Sample3.txt", 5L),
                Arguments.of("day09Sample4.txt", 2L)
            )
    }
}

private fun Long.shouldBe(expectation: Long) = assertEquals(expectation, this)
